mkosi/distributions/gentoo.py

```diff
@@ -1,198 +1,99 @@
+from mkosi.types import PathString
 
 
 def invoke_emerge(
     state: MkosiState,
+    sysroot: Optional[Path] = None,
+    bwrap_params: list[PathString] = [],
     pkgs: Sequence[str] = (),
     actions: Sequence[str] = (),
     opts: Sequence[str] = (),
+    env: dict[str, str] = {},
 ) -> None:
-    jobs = os.cpu_count() or 1
+    thread_counts = (os.cpu_count() or 1) * 2  # * 2 for hyperthreading
+    bwrap: list[PathString] = []
+    root: Optional[Path] = None
+    if sysroot is not None:
+        # This is the mount-point inside our sysroot where we mount root
+        target_root_mntp = "/tmp/mkosi-root"
+        bwrap += ["--bind", state.root, target_root_mntp]
+        root = Path(target_root_mntp)
+    else:
+        sysroot = state.root
+
     emerge_default_opts = [
         "--buildpkg=y",
         "--usepkg=y",
         "--keep-going=y",
-        f"--jobs={jobs}",
-        f"--load-average={jobs+1}",
+        f"--jobs={thread_counts}",
+        f"--load-average={thread_counts+1}",
         "--nospinner",
+        *([f"--root={root}"] if root else []),
     ]
     if ARG_DEBUG.get():
         emerge_default_opts += ["--verbose", "--quiet=n", "--quiet-fail=n"]
     else:
         emerge_default_opts += ["--quiet-build", "--quiet"]
     cmd = ["emerge", *pkgs, *emerge_default_opts, *opts, *actions]
-    run_workspace_command(state.root, cmd, network=True, env=state.environment)
-
-
-class Gentoo:
-    arch_profile: Path
-    arch: str
-    custom_profile_path: Path
-    ebuild_sh_env_dir: Path
-    emerge_vars: dict[str, str]
-    portage_cfg_dir: Path
-    profile_path: Path
-    root: Path
-    pkgs: dict[str, list[str]] = {}
-
-    EMERGE_UPDATE_OPTS = [
-        "--update",
-        "--tree",
-        "--changed-use",
-        "--newuse",
-        "--deep",
-        "--with-bdeps=y",
-        "--complete-graph-if-new-use=y",
-        "--verbose-conflicts",
+    bwrap += [
+        "--bind", state.cache_dir / "binpkgs", "/var/cache/binpkgs",
+        "--bind", state.cache_dir / "distfiles", "/var/cache/distfiles",
+        "--bind", state.cache_dir / "repos", "/var/db/repos",
+        *bwrap_params
     ]
+    run_workspace_command(sysroot, cmd, bwrap_params=bwrap, network=True,
+                          env=env)
 
-    portage_use_flags = [
-        "initramfs",
-        "git",  # for sync-type=git
-        "symlink",  # for kernel
-    ]
 
-    # TODO: portage_features.add("ccache"), this shall expedite the builds
-    portage_features = [
-        # -user* are required for access to USER_CONFIG_PATH
-        "-userfetch",
-        "-userpriv",
-        "-usersync",
-        "-usersandbox",
-        "-sandbox",
-        "-pid-sandbox",  # for cross-compile scenarios
-        "-network-sandbox",
-        "parallel-install",
-        "getbinpkg",
-        "-candy",
-    ]
+class GentooInstaller(DistributionInstaller):
+    stage3_cache: Path
+
+    @classmethod
+    def filesystem(cls) -> str:
+        return "btrfs"
```

Comment:
Please add some justification to the commit message for this change? What makes btrfs a better default filesystem than ext4 for Gentoo?

Reply:
i will update the commit msg with i just said above. let me know.

Reply:
done!